File: src/expansion/scala/mrtjp/projectred/expansion/block/ElectrotineGeneratorBlock.scala
package mrtjp.projectred.expansion.block

import mrtjp.projectred.expansion.{BaseMachineBlock, RotatableMachineBlock, TileElectrotineGenerator}
import net.minecraft.block.{Block, BlockState}
import net.minecraft.item.BlockItemUseContext
import net.minecraft.state.StateContainer

class ElectrotineGeneratorBlock extends RotatableMachineBlock(() => new TileElectrotineGenerator) {

    override protected def createBlockStateDefinition(builder: StateContainer.Builder[Block, BlockState]): Unit = {
        super.createBlockStateDefinition(builder)
        builder.add(BaseMachineBlock.CHARGED_PROPERTY)
        builder.add(BaseMachineBlock.WORKING_PROPERTY)
    }

    override def getStateForPlacement(context: BlockItemUseContext): BlockState = {
        super.getStateForPlacement(context)
            .setValue(BaseMachineBlock.CHARGED_PROPERTY, Boolean.box(false))
            .setValue(BaseMachineBlock.WORKING_PROPERTY, Boolean.box(false))
    }
}

File: core/src/main/java/mrtjp/projectred/core/block/ElectrotineGeneratorBlock.java
package mrtjp.projectred.core.block;

import codechicken.lib.vec.Rotation;
import mrtjp.projectred.core.init.CoreBlocks;
import mrtjp.projectred.core.tile.ElectrotineGeneratorBlockEntity;
import net.minecraft.core.BlockPos;
import net.minecraft.world.item.context.BlockPlaceContext;
import net.minecraft.world.level.block.Block;
import net.minecraft.world.level.block.entity.BlockEntity;
import net.minecraft.world.level.block.entity.BlockEntityType;
import net.minecraft.world.level.block.state.BlockState;
import net.minecraft.world.level.block.state.StateDefinition;

import javax.annotation.Nullable;

public class ElectrotineGeneratorBlock extends ProjectRedBlock {

    public ElectrotineGeneratorBlock() {
        super(STONE_PROPERTIES);
    }

    @Nullable
    @Override
    public BlockEntity newBlockEntity(BlockPos pos, BlockState state) {
        return new ElectrotineGeneratorBlockEntity(pos, state);
    }

    @Override
    protected BlockEntityType<?> getBlockEntityType() {
        return CoreBlocks.ELECTROTINE_GENERATOR_BLOCK_ENTITY.get();
    }

    @Nullable
    @Override
    public BlockState getStateForPlacement(BlockPlaceContext context) {
        int r = Rotation.rotationTo(0, context.getHorizontalDirection().ordinal());
        return this.defaultBlockState()
                .setValue(ROTATION, r)
                .setValue(CHARGED, false)
                .setValue(WORKING, false);
    }

    @Override
    protected void createBlockStateDefinition(StateDefinition.Builder<Block, BlockState> builder) {
        builder.add(ROTATION);
        builder.add(CHARGED);
        builder.add(WORKING);
    }
}
